### DISCUSSION TITLE
Ubuntu18.04 cant install rustdesk properly

### DISCUSSION THREAD

**Discussion by Julian-Won** [rustdesk/rustdesk#6148] [General]
### Bug Description

When I install the latest rustdesk on ubuntu 18.04,it directly display an error message:
```
test@test-Standard-PC-i440FX-PIIX-1996:~$ sudo apt install -fy ./rustdesk-1.2.3-x86_64.deb 
Reading package lists... Done
Building dependency tree       
Reading state information... Done
Note, selecting 'rustdesk' instead of './rustdesk-1.2.3-x86_64.deb'
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 rustdesk : Depends: gstreamer1.0-pipewire but it is not installable

```
then I used apt software to install gstreamer ,it cant fond the gstreamer1.0-pipewire

### How to Reproduce

Firstly use Ubuntu18.04 to 
`wget https://github.com/rustdesk/rustdesk/releases/download/1.2.3/rustdesk-1.2.3-x86_64.deb`
and then use prompt
`sudo apt install -fy ./rustdesk-1.2.3-x86_64.deb`
then , it shows this error
```
test@test-Standard-PC-i440FX-PIIX-1996:~$ sudo apt install -fy ./rustdesk-1.2.3-x86_64.deb 
Reading package lists... Done
Building dependency tree       
Reading state information... Done
Note, selecting 'rustdesk' instead of './rustdesk-1.2.3-x86_64.deb'
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 rustdesk : Depends: gstreamer1.0-pipewire but it is not installable
E: Unable to correct problems, you have held broken packages.
```

### Expected Behavior

Install the rustdesk on Ubuntu18.04 successfully

### Operating system(s) on local side and remote side

Ubuntu 20.04

### RustDesk Version(s) on local side and remote side

1.2.3

### Screenshots

![image](https://github.com/rustdesk/rustdesk/assets/67736924/aebc2ca8-9f27-491e-a499-c4d39d773d53)

**Comment by paulsommer** [comment]
According to packages.ubuntu.com this package is only available for Ubuntu 22.04, 22.10, 23.04 and 23.10
https://packages.ubuntu.com/search?keywords=gstreamer1.0-pipewire&searchon=names&suite=all&section=allMy RustDesk Version 1.2.0 doesn't need that package while 1.2.3 does.

Either you should switch to Ubuntu >= 22 or use Rustdesk <= 1.2.0.
I would strongly suggest to upgrade Ubuntu as 18.04 is already EOL (End Of Life ).

**Reply by Julian-Won** [reply]
Thanks for your reply, and i will replace my software version to install on 18.04 which is strongly compatible with some work sortware! Thanks angin ><)!

**Comment by rustdesk** [comment]
Please add this source http://ppa.launchpad.net/pipewire-debian/pipewire-upstream/ubuntu for gstreamer1.0-pipewire in Ubuntu as 18.04

https://pipewire-debian.github.io/pipewire-debian/

```
sudo apt install software-properties-common
sudo add-apt-repository ppa:pipewire-debian/pipewire-upstream
sudo apt update
```

**Reply by mmallejac** [reply]
Tried that as well, no more luck on my 18.04 box.

Yes `gstreamer1.0-pipewire:amd64` is installed, but I now have the following error while starting rustdesk :
`Failed to load "librustdesk.so"
/usr/lib/support4next/lib/librustdesk.so: undefined symbol: fcntl64`

Maybe, as suggested [here](https://github.com/bincrafters/community/issues/962), rustdesk should be built in another way ?

**Reply by DeadZen** [reply]
https://linuxconfig.org/how-to-install-pipewire-on-ubuntu-linux
then dpkg -i theoneforyoursystem.deb 
.. that worked for me

**Reply by Myrkie** [reply]
also getting this issue though with a different path
<details>

<summary>code</summary>

myrkur@TessaCyan:~$ rustdesk

Failed to load "librustdesk.so"

/usr/share/rustdesk/lib/librustdesk.so: undefined symbol: fcntl64

</details>

**Comment by Den41k92** [comment]
This is worked for me on Kubuntu 24.04 :
Download a .deb image for linux from releases page
```sh
sudo apt remove rustdesk
sudo add-apt-repository ppa:pipewire-debian/pipewire-upstream
sudo apt install -y pipewire gstreamer1.0-pipewire libxdo3
sudo dpkg -i rustdesk-1.3.8-x86_64.deb
```

**Reply by MyUserNameIsMy** [reply]
I faced same issue ```Failed to load "librustdesk.so" /usr/lib/support4next/lib/librustdesk.so: undefined symbol: fcntl64```
I tried 1.4 and 1.3.8 versions using @Den41k92 solution, however it keeps logging error above.

**Comment by laygond** [comment]
Whenever a debian package fails consider installing the flatpack or appimage version. I installed the Rustdesk AppImage 64 bit for Ubuntu18.04 with no issues from their github [assets releases](https://github.com/rustdesk/rustdesk/releases/tag/1.4.0)

**Comment by sayoUNL** [comment]
In case you can find fcntl (but not fcntl64) in your system:
```
#include <fcntl.h>
#include <stdarg.h>

int fcntl64(int fd, int cmd, ...) {
    va_list args;
    va_start(args, cmd);
    int ret;

    if (cmd == F_SETLK || cmd == F_SETLKW || cmd == F_GETLK) {
        struct flock *lock = va_arg(args, struct flock *);
        ret = fcntl(fd, cmd, lock);
    } else {
        long arg = va_arg(args, long);
        ret = fcntl(fd, cmd, arg);
    }

    va_end(args);
    return ret;
}
```
You can compile the code to get libfcntl64.so:

`gcc -shared -fPIC fcntl64.c -o libfcntl64.so`

Then you can turn the app on by Executing:

`LD_PRELOAD=/path/to/libfcntl64.so rustdesk`
 (change `/path/to/` to where libfcntl64.so actually is, don't move libfcntl64.so to another place)
